File: app/src/main/java/vn/edu/utt/uttqlsv/view/fragments/EditStudent.java
package vn.edu.utt.uttqlsv.view.fragments;


import android.annotation.SuppressLint;
import android.app.DatePickerDialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

import vn.edu.utt.uttqlsv.R;
import vn.edu.utt.uttqlsv.controller.StudentListManager;
import vn.edu.utt.uttqlsv.controller.utils.Validator;
import vn.edu.utt.uttqlsv.model.Gender;
import vn.edu.utt.uttqlsv.model.Student;
import vn.edu.utt.uttqlsv.view.activities.StudentList;

public class EditStudent extends Fragment {

    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_STUDENT = "student_param";
    int initYear;
    int initMonth;
    int initDate;
    private Student student;

    public static EditStudent newInstance(String studentCode) {
        EditStudent fragment = new EditStudent();
        Bundle bundle = new Bundle();
        bundle.putString(ARG_STUDENT, studentCode);
        fragment.setArguments(bundle);
        return fragment;
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getArguments() != null) {
            String studentCode = getArguments().getString(ARG_STUDENT);

            for (Student s : StudentListManager.getInstance().getStudentList()) {
                if (studentCode.equals(s.getStudentCode())) {
                    student = s;
                }
            }
        }
    }

    @SuppressLint("SetTextI18n")
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_edit_student, container, false);

        Spinner spinner = v.findViewById(R.id.edit_student_fragment_gender_spinner);
        String[] items = {"Male", "Female"};
        ArrayAdapter adapter = new ArrayAdapter(requireContext(), android.R.layout.simple_spinner_dropdown_item, items);
        spinner.setAdapter(adapter);

        if (student.isGender() == Gender.MALE) spinner.setSelection(0);
        else spinner.setSelection(1);

        //Init view for each property of student
        TextView studentCodeEdt = v.findViewById(R.id.edit_student_fragment_student_code_edt);
        EditText nameEdt = v.findViewById(R.id.edit_student_fragment_student_name_edt);
        EditText classNameEdt = v.findViewById(R.id.edit_student_fragment_class_name_edt);
        EditText gradeEdt = v.findViewById(R.id.edit_student_fragment_grade_edt);
        EditText addressEdt = v.findViewById(R.id.edit_student_fragment_add_student_address_edt);
        EditText mathEdt = v.findViewById(R.id.edit_student_fragment_math_score_edt);
        EditText literatureEdt = v.findViewById(R.id.edit_student_fragment_literature_score_edt);
        EditText englishEdt = v.findViewById(R.id.edit_student_fragment_english_score_edt);
        EditText physicEdt = v.findViewById(R.id.edit_student_fragment_physic_score_edt);
        EditText geographyEdt = v.findViewById(R.id.edit_student_fragment_geography_score_edt);
        EditText historyEdt = v.findViewById(R.id.edit_student_fragment_history_score_edt);
        EditText biologyEdt = v.findViewById(R.id.edit_student_fragment_biology_score_edt);
        EditText chemistryEdt = v.findViewById(R.id.edit_student_fragment_chemistry_score_edt);
        Button saveBtn = v.findViewById(R.id.save_btn);
        TextView dobTV = v.findViewById(R.id.edit_student_fragment_date_of_birth_tv);
        ImageButton datePickerBtn = v.findViewById(R.id.edit_student_fragment_show_date_picker);


        try {
            @SuppressLint("SimpleDateFormat") Date date = new SimpleDateFormat("dd/MM/yyyy").parse(student.getDateOfBirth());
            if (date == null) date = new Date(System.currentTimeMillis());

            initYear = date.getYear() + 1900;
            initMonth = date.getMonth() + 1;
            initDate = date.getDate();

            dobTV.setText(initDate + "/" + initMonth + "/" + initYear);
            DatePickerDialog.OnDateSetListener dateSetListener = (view, year, month, dayOfMonth) -> {
                dobTV.setText(dayOfMonth + "/" + (month + 1) + "/" + year);
                initYear = year;
                initMonth = month + 1;
                initDate = dayOfMonth;
            };

            DatePickerDialog dialog = new DatePickerDialog(requireContext(), android.R.style.Theme_Material_Light_Dialog_NoActionBar, dateSetListener, initYear, initMonth, initYear);
            datePickerBtn.setOnClickListener(v1 -> dialog.show());

            //set initial value for each view of properties
            studentCodeEdt.setText(student.getStudentCode());
            nameEdt.setText(student.getName());
            classNameEdt.setText(student.getClassName());
            gradeEdt.setText(student.getGrade());
            addressEdt.setText(student.getAddress());
            mathEdt.setText(String.valueOf(student.getMathScore()));
            literatureEdt.setText(String.valueOf(student.getLiteratureScore()));
            englishEdt.setText(String.valueOf(student.getEnglishScore()));
            physicEdt.setText(String.valueOf(student.getPhysicScore()));
            geographyEdt.setText(String.valueOf(student.getGeographyScore()));
            historyEdt.setText(String.valueOf(student.getHistoryScore()));
            chemistryEdt.setText(String.valueOf(student.getChemistryScore()));
            biologyEdt.setText(String.valueOf(student.getBiologyScore()));

            //Handle save button click event
            saveBtn.setOnClickListener(v1 -> {
                String studentName = nameEdt.getText().toString().trim();
                String className = classNameEdt.getText().toString().trim();
                String grade = gradeEdt.getText().toString().trim();
                boolean gender = spinner.getSelectedItemId() == 0 ? Gender.MALE : Gender.FEMALE;
                String address = addressEdt.getText().toString().trim();
                String mathStr = mathEdt.getText().toString().trim();
                String literatureStr = literatureEdt.getText().toString().trim();
                String englishStr = englishEdt.getText().toString().trim();
                String physicStr = physicEdt.getText().toString().trim();
                String geographyStr = geographyEdt.getText().toString().trim();
                String historyStr = historyEdt.getText().toString().trim();
                String chemistryStr = chemistryEdt.getText().toString().trim();
                String biologyStr = biologyEdt.getText().toString().trim();

                Student newStudent = edit(studentName, className, grade, gender, address, mathStr, literatureStr, englishStr, physicStr, geographyStr, historyStr, chemistryStr, biologyStr);
                if (newStudent != null) {
                    StudentListManager.getInstance().updateStudentInfo(student, newStudent);
                }

                StudentList studentListActivity = (StudentList) getActivity();
                studentListActivity.onBackPressed();
                studentListActivity.refreshStudentList();

            });

        } catch (ParseException e) {
            e.printStackTrace();
        }

        return v;
    }

    private Student edit(String studentName, String className, String grade, boolean gender, String address, String mathStr, String literatureStr, String englishStr, String physicStr, String geographyStr, String historyStr, String chemistryStr, String biologyStr) {
        Student student = new Student();
        student.setName(studentName);
        student.setClassName(className);
        if (!grade.isEmpty()) student.setGrade(Integer.parseInt(grade));
        else student.setGrade(10);

        student.setGender(gender);
        student.setAddress(address);

        if (!mathStr.isEmpty() && Validator.isValidFloatingNumber(mathStr))
            student.setMathScore(Float.parseFloat(mathStr));
        else student.setMathScore(0f);

        if (!literatureStr.isEmpty() && Validator.isValidFloatingNumber(literatureStr))
            student.setLiteratureScore(Float.parseFloat(literatureStr));
        else student.setLiteratureScore(0f);

        if (!englishStr.isEmpty() && Validator.isValidFloatingNumber(englishStr))
            student.setEnglishScore(Float.parseFloat(englishStr));
        else student.setEnglishScore(0f);

        if (!physicStr.isEmpty() && Validator.isValidFloatingNumber(physicStr))
            student.setPhysicScore(Float.parseFloat(physicStr));
        else student.setPhysicScore(0f);

        if (!geographyStr.isEmpty() && Validator.isValidFloatingNumber(geographyStr))
            student.setGeographyScore(Float.parseFloat(geographyStr));
        else student.setGeographyScore(0f);

        if (!historyStr.isEmpty() && Validator.isValidFloatingNumber(historyStr))
            student.setHistoryScore(Float.parseFloat(historyStr));
        else student.setHistoryScore(0f);

        if (!chemistryStr.isEmpty() && Validator.isValidFloatingNumber(chemistryStr))
            student.setChemistryScore(Float.parseFloat(chemistryStr));
        else student.setChemistryScore(0f);

        if (!biologyStr.isEmpty() && Validator.isValidFloatingNumber(biologyStr))
            student.setBiologyScore(Float.parseFloat(biologyStr));
        else student.setBiologyScore(0f);

        String dobStr = toDateString(initDate, initMonth, initYear);

        if (Validator.isValidDateString(dobStr)) {
            student.setDateOfBirth(dobStr);
            return student;
        }

        return null;
    }

    private String toDateString(int day, int month, int year) {
        String s = "";
        s += (day < 10) ? ("0" + day + "/") : (day + "/");
        s += (month < 10) ? ("0" + month + "/") : (month + "/");
        s += year;
        return s;
    }
}

File: app/src/main/java/vn/edu/utt/uttqlsv/view/fragments/EditStudent.kt
package vn.edu.utt.uttqlsv.view.fragments

import android.annotation.SuppressLint
import android.app.DatePickerDialog
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import kotlinx.android.synthetic.main.activity_add_student.*
import kotlinx.android.synthetic.main.activity_sign_up.*
import kotlinx.android.synthetic.main.fragment_edit_student.*
import vn.edu.utt.uttqlsv.R
import vn.edu.utt.uttqlsv.controller.StudentListManager
import vn.edu.utt.uttqlsv.controller.utils.Validator
import vn.edu.utt.uttqlsv.model.Gender
import vn.edu.utt.uttqlsv.model.Student
import vn.edu.utt.uttqlsv.view.activities.StudentList
import java.text.SimpleDateFormat
import java.util.*

// TODO: Rename parameter arguments, choose names that match
// the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
private const val ARG_STUDENT = "student_param"

/**
 * A simple [Fragment] subclass.
 * Use the [EditStudent.newInstance] factory method to
 * create an instance of this fragment.
 */
@Suppress("DEPRECATION")
class EditStudent : Fragment() {
    // TODO: Rename and change types of parameters
    private var studentCode: String? = null
    private var student: Student? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            studentCode = it.getString(ARG_STUDENT)

            for (a in StudentListManager.studentList) {
                if (studentCode.equals(a.studentCode)) {
                    student = a
                }
            }
        }
    }

    @SuppressLint("MissingInflatedId", "SimpleDateFormat")
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        val v = inflater.inflate(R.layout.fragment_edit_student, container, false)

        //init spinner for gender picker dropdown menu
        val genderDropDownListItems = arrayOf("Male", "Female")
        val adapter = ArrayAdapter(
            requireContext(), android.R.layout.simple_spinner_item, genderDropDownListItems
        )

        val spinner: Spinner = v.findViewById(R.id.edit_student_fragment_gender_spinner)
        spinner.adapter = adapter

        if (student?.gender == Gender.MALE) spinner.setSelection(0)
        else spinner.setSelection(1)

        //Init view for each property of Student
        val studentCodeEdt: TextView = v.findViewById(R.id.edit_student_fragment_student_code_edt)
        val nameEdt: EditText = v.findViewById(R.id.edit_student_fragment_student_name_edt)
        val classNameEdt: EditText = v.findViewById(R.id.edit_student_fragment_class_name_edt)
        val gradeEdt: EditText = v.findViewById(R.id.edit_student_fragment_grade_edt)
        val addressEdt: EditText =
            v.findViewById(R.id.edit_student_fragment_add_student_address_edt)
        val mathEdt: EditText = v.findViewById(R.id.edit_student_fragment_math_score_edt)
        val literatureEdt: EditText =
            v.findViewById(R.id.edit_student_fragment_literature_score_edt)
        val englishEdt: EditText = v.findViewById(R.id.edit_student_fragment_english_score_edt)
        val physicEdt: EditText = v.findViewById(R.id.edit_student_fragment_physic_score_edt)
        val geographyEdt: EditText = v.findViewById(R.id.edit_student_fragment_geography_score_edt)
        val historyEdt: EditText = v.findViewById(R.id.edit_student_fragment_history_score_edt)
        val biologyEdt: EditText = v.findViewById(R.id.edit_student_fragment_biology_score_edt)
        val chemistryEdt: EditText = v.findViewById(R.id.edit_student_fragment_chemistry_score_edt)
        val saveBtn: Button = v.findViewById(R.id.save_btn)
        val dobTV: TextView = v.findViewById(R.id.edit_student_fragment_date_of_birth_tv)
        val datePickerBtn: ImageButton = v.findViewById(R.id.edit_student_fragment_show_date_picker)


        var date = SimpleDateFormat("dd/MM/yyyy").parse(student?.dateOfBirth!!)
        if (date == null) date = Date(System.currentTimeMillis())
        Log.d("Edit Student", "$date")

        var initYear = date.year + 1900
        var initMonth = date.month + 1
        var initDate = date.date

        dobTV.text = "$initDate-$initMonth-$initYear"
        val dateSetListener = DatePickerDialog.OnDateSetListener { _, year, month, dayOfMonth ->
            dobTV.text = "$dayOfMonth/${month + 1}/$year"
            initYear = year
            initMonth = month + 1
            initDate = dayOfMonth
        }
        val datePickerDialog = DatePickerDialog(
            requireContext(),
            android.R.style.Theme_Material_Light_Dialog_NoActionBar,
            dateSetListener,
            initYear,
            initMonth,
            initDate
        )

        datePickerBtn.setOnClickListener { datePickerDialog.show() }

        //set initial value for each view of properties
        studentCodeEdt.text = student?.studentCode
        nameEdt.setText(student?.name)
        classNameEdt.setText(student?.className)
        gradeEdt.setText(student?.grade.toString())
        addressEdt.setText(student?.address)
        mathEdt.setText(student?.mathScore.toString())
        literatureEdt.setText(student?.literatureScore.toString())
        englishEdt.setText(student?.englishScore.toString())
        physicEdt.setText(student?.physicScore.toString())
        geographyEdt.setText(student?.geographyScore.toString())
        historyEdt.setText(student?.historyScore.toString())
        chemistryEdt.setText(student?.chemistryScore.toString())
        biologyEdt.setText(student?.biologyScore.toString())

        //Handle Save Button click event
        saveBtn.setOnClickListener {

            val studentName = nameEdt.text.toString().trim()
            val className = classNameEdt.text.toString().trim()
            val grade = gradeEdt.text.toString().trim()
            val gender: Boolean =
                if (edit_student_fragment_gender_spinner.selectedItemId == "0".toLong()) Gender.MALE
                else Gender.FEMALE

            val address = addressEdt.text.toString().trim()
            val mathScoreStr = mathEdt.text.toString().trim()
            val literatureScoreStr = literatureEdt.text.toString().trim()
            val englishScoreStr = englishEdt.text.toString().trim()
            val physicScoreStr = physicEdt.text.toString().trim()
            val geographyScoreStr = geographyEdt.text.toString().trim()
            val historyScoreStr = historyEdt.text.toString().trim()
            val chemistryScoreStr = chemistryEdt.text.toString().trim()
            val biologyScoreStr = biologyEdt.text.toString().trim()

            val newStudent = Student()
            newStudent.name = studentName
            newStudent.className = className
            if (grade.isNotEmpty())
                newStudent.grade = Integer.parseInt(grade)
            else
                newStudent.grade = 10

            newStudent.gender = gender
            newStudent.address = address

            if (mathScoreStr.isNotEmpty() && Validator.isValidFloatingNumber(mathScoreStr)) newStudent.mathScore =
                mathScoreStr.toFloat()
            else newStudent.mathScore = 0f

            if (literatureScoreStr.isNotEmpty() && Validator.isValidFloatingNumber(
                    literatureScoreStr
                )
            ) newStudent.literatureScore = literatureScoreStr.toFloat()
            else newStudent.literatureScore = 0f

            if (englishScoreStr.isNotEmpty() && Validator.isValidFloatingNumber(englishScoreStr)) newStudent.englishScore =
                englishScoreStr.toFloat()
            else newStudent.englishScore = 0f

            if (physicScoreStr.isNotEmpty() && Validator.isValidFloatingNumber(physicScoreStr)) newStudent.physicScore =
                physicScoreStr.toFloat()
            else newStudent.physicScore = 0f

            if (geographyScoreStr.isNotEmpty() && Validator.isValidFloatingNumber(geographyScoreStr)) newStudent.geographyScore =
                geographyScoreStr.toFloat()
            else newStudent.geographyScore = 0f

            if (historyScoreStr.isNotEmpty() && Validator.isValidFloatingNumber(historyScoreStr)) newStudent.historyScore =
                historyScoreStr.toFloat()
            else newStudent.historyScore = 0f

            if (chemistryScoreStr.isNotEmpty() && Validator.isValidFloatingNumber(chemistryScoreStr)) newStudent.chemistryScore =
                chemistryScoreStr.toFloat()
            else newStudent.chemistryScore = 0f

            if (biologyScoreStr.isNotEmpty() && Validator.isValidFloatingNumber(biologyScoreStr)) newStudent.biologyScore =
                biologyScoreStr.toFloat()
            else newStudent.biologyScore = 0f

            val dobStr = toDateString(initDate, initMonth, initYear)
            Log.d("Add Student", "Date String $dobStr")
            if (Validator.isValidDateString(dobStr)) {
                newStudent.dateOfBirth = dobStr
                StudentListManager.updateStudentInfo(student!!, newStudent)
            }


            val studentListActivity = activity as? StudentList
            studentListActivity?.onBackPressed()
            studentListActivity?.refreshStudentList()
        }

        return v
    }


    private fun toDateString(day: Int, month: Int, year: Int): String {
        var s = ""
        s += if (day < 10) "0$day/"
        else "$day/"

        s += if (month < 10) "0$month/"
        else "$month/"
        s += "$year"
        return s
    }

    companion object {
        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @return A new instance of fragment EditStudent.
         */
        // TODO: Rename and change types and number of parameters
        @JvmStatic
        fun newInstance(studentCode: String) = EditStudent().apply {
            arguments = Bundle().apply {
                putString(ARG_STUDENT, studentCode)
            }
        }
    }

}